vendor/bat-native-ledger/src/bat/ledger/internal/core/async_result.h

```diff
@@ -0,0 +1,122 @@
+//   resolver.Complete(42);
+//
+//   AsyncResult<int> result = resolver.result();
+//   result.Listen(base::BindOnce([](const int& value) {}));
+//
+// Listeners are called on the current SequencedTaskRunner, and are guaranteed
+// to be called asynchronously. AsyncResult and Resolver objects are internally
+// reference counted and can be passed between sequences; the internal data
+// structures are updated on the sequence that created the Resolver.
+template <typename T>
+class AsyncResult {
+ public:
+  using CompleteType = T;
+  using CompleteCallback = base::OnceCallback<void(const T&)>;
+
+  void Listen(CompleteCallback on_complete) {
+    Listener listener = {.on_complete = std::move(on_complete),
+                         .task_runner = base::SequencedTaskRunnerHandle::Get()};
+
+    task_runner_->PostTask(FROM_HERE, base::BindOnce(AddListenerInTask, store_,
+                                                     std::move(listener)));
+  }
+
+  class Resolver {
+   public:
+    Resolver() {}
+    void Complete(T&& value) { result_.Complete(std::move(value)); }
+    AsyncResult result() const { return result_; }
+
+   private:
+    AsyncResult result_;
+  };
+
+ private:
+  AsyncResult()
+      : store_(new Store()),
+        task_runner_(base::SequencedTaskRunnerHandle::Get()) {}
+
+  enum class State { kPending, kComplete };
+
+  struct Listener {
+    CompleteCallback on_complete;
+    scoped_refptr<base::SequencedTaskRunner> task_runner;
+  };
+
+  struct Store {
+    Store() {}
+    State state = State::kPending;
+    base::Optional<T> value;
+    std::list<Listener> listeners;
+  };
+
+  void Complete(T&& value) {
+    task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(SetCompleteInTask, store_, std::move(value)));
+  }
+
+  static void AddListenerInTask(std::shared_ptr<Store> store,
+                                Listener listener) {
+    switch (store->state) {
+      case State::kComplete:
+        listener.task_runner->PostTask(
+            FROM_HERE, base::BindOnce(RunCompleteCallback, store,
+                                      std::move(listener.on_complete)));
+        break;
+      case State::kPending:
+        store->listeners.emplace_back(std::move(listener));
+        break;
+    }
+  }
+
+  static void SetCompleteInTask(std::shared_ptr<Store> store, T value) {
+    if (store->state != State::kPending)
+      return;
+
+    store->state = State::kComplete;
+    store->value = std::move(value);
+
+    for (auto& listener : store->listeners) {
+      listener.task_runner->PostTask(
+          FROM_HERE, base::BindOnce(RunCompleteCallback, store,
+                                    std::move(listener.on_complete)));
+    }
+
+    store->listeners.clear();
+  }
+
+  static void RunCompleteCallback(std::shared_ptr<Store> store,
+                                  CompleteCallback on_complete) {
+    DCHECK(store->value);
+    std::move(on_complete).Run(*store->value);
+  }
+
+  std::shared_ptr<Store> store_;
```

Comment:
shared_ptr is a banned feature https://chromium-cpp.appspot.com/

Reply:
I'm aware, and since this is an internal use it could be changed to use `base::RefCounted` without changing the interface. A couple of considerations that lead to using `std::shared_ptr` here:

- `std::shared_ptr` is used throughout the `bat-native-ledger` code.
- It really is better than `base::RefCounted` for this use case.
- It really is better than `base::RefCounted` for this use case.